de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/concurrent/management/JupiterClient.java

```diff
@@ -7,34 +7,68 @@
 import de.fu_berlin.inf.dpp.concurrent.jupiter.Operation;
 import de.fu_berlin.inf.dpp.concurrent.jupiter.TransformationException;
 import de.fu_berlin.inf.dpp.concurrent.jupiter.internal.Jupiter;
+import de.fu_berlin.inf.dpp.concurrent.jupiter.internal.text.TimestampOperation;
+import de.fu_berlin.inf.dpp.session.AbstractActivityProducer;
 import de.fu_berlin.inf.dpp.session.ISarosSession;
+import de.fu_berlin.inf.dpp.util.ThreadUtils;
 import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.log4j.Logger;
 
 /** A JupiterClient manages Jupiter client docs for a single user with several paths */
-public class JupiterClient {
+public class JupiterClient extends AbstractActivityProducer {
 
   protected ISarosSession sarosSession;
-
-  public JupiterClient(ISarosSession sarosSession) {
-    this.sarosSession = sarosSession;
-  }
-
+  private static Logger log = Logger.getLogger(JupiterClient.class);
   /**
    * Jupiter instances for each local editor.
    *
    * @host and @client
    */
   protected final HashMap<SPath, Jupiter> clientDocs = new HashMap<SPath, Jupiter>();
 
+  private final AtomicBoolean stopHeartbeatThread = new AtomicBoolean(false);
+  private final Thread jupiterHeartbeat =
+      ThreadUtils.runSafeAsync(
+          "dpp-jupiter-heartbeat",
+          log,
+          () -> {
+            while (!stopHeartbeatThread.get()) {
```

Comment:
according to "Java Concurrency in Practice - 7.1.1 Interruption" think about using interruption for cancellation 
`while (!Thread.currentThread().isInterrupted())`